xcoll/lossmap.py

```diff
@@ -1,44 +1,175 @@
 # copyright ############################### #
 # This file is part of the Xcoll Package.   #
-# Copyright (c) CERN, 2023.                 #
+# Copyright (c) CERN, 2024.                 #
 # ######################################### #
+import numpy as np
+import pandas as pd
+from pathlib import Path
+import json
 
+import xtrack as xt
+import xpart as xp
+import xobjects as xo
 
+from .beam_elements import _all_collimator_types
 
-
+# TODO: should we add a check to make sure line_is_reversed is a boolean/correct value?
 
 class LossMap:
-    
-    def __init__(self, line, part, line_is_reversed):
+
+    def __init__(self, line, line_is_reversed, part):
         self._line = line
         self._line_is_reversed = line_is_reversed
         self._machine_length = line.get_length()
-        self._lossmap = None
+        self._lossmap = None                            
         self._summary = None
+        self._part = part
+
+    # TODO: Check if _all_collimator_types is valid
+   
+    def lossmap(self, interpolation = 0.1, weights = None):
+        """ 
+            Does the loss location refinement and interpolation. 
+        """
+        # loss location refinement
+        if interpolation is not None: 
+            new_state = self._part.state                   
+            new_elem  = self._part.at_element               
+
+            for idx, elem in enumerate(self._part.at_element):
+                if (self._part.state[idx] == 0 and elem > 0 
+                    and self._line.element_names[elem - 1] in 
+                    self._line.get_elements_of_type(_all_collimator_types)[1]):
+
+                    print(f"Found at {self._line.element_names[elem]}, should be {self._line.element_names[elem-1]}")
+                    new_elem[idx] = elem - 1
+                    what_type = self._line.element_names[elem-1].__class__.__name__
+                    if what_type   == 'EverestCollimator':
```

Comment:
Need to make this more general, based on particle states inside particle_states.h (open TODO for later)